### PR TITLE
MPR#7290: ocamldoc, more support for inline record

### PR DESCRIPTION
**Pull request by Octachron**
This pull request improves the support of inline record within ocamldoc:
- First, all code path should be implemented and using inline record should not
  lead anymore to ocamldoc crashes
- Second, the PR adds a light support for documentation comments within inline record.
  However, the backend support for these documentation comments could be improved.

Note that the first commit in this PR is a questionable refactoring commit that I implemented when trying to decrease my perception of the signal to noise ratio in `odoc_sig.ml`, it can be reverted.

(see also the associated [mantis ticket](http://caml.inria.fr/mantis/view.php?id=7290))

### CODE REVIEW

File: testsuite/tests/tool-ocamldoc-2/inline_records.mli [hunk 1]
```diff
@@ -1,14 +1,34 @@
...
-type r = {lbl: int (** Field documentation *);
-          more:int list (** More documentation *) }
+type r = { lbl: int (** Field documentation for non-inline *);
+          more:int list (** More documentation for r *) }
+
+
+(** A sum type with one inline record *)
+type t = A of {lbl: int (** A field documentation *); more:int list (** More A field documentation *) }  (** Constructor documentation *)
+
+(** A sum type with two inline records *)
```
Comment (gasche) [inline testsuite/tests/tool-ocamldoc-2/inline_records.mli:26]: Could we also have a GADT with inline record in the testsuite?

File: testsuite/tests/tool-ocamldoc-2/inline_records.reference [hunk 2]
```diff
@@ -50,14 +76,114 @@ A simple record type for reference
 
 \label{TYPInline-underscorerecords.t}\begin{ocamldoccode}
 type t =
-  | C of {\char123}
-   lbl : int;
-   more : int list;
-{\char125}
+
 \end{ocamldoccode}
```
Comment (gasche) [inline testsuite/tests/tool-ocamldoc-2/inline_records.reference:102]: Do you know why this environment now ends before the code output?

Comment (gasche) [inline testsuite/tests/tool-ocamldoc-2/inline_records.reference:102]: It also looks like the field documentations are actually not processed. Is that a mistake on my part reading the output?

**Comment by gasche** [comment]
This looks very nice, thanks. I'm a bit surprised by the bracketing of code elements in the test outputs (see inline comment for LaTeX). Did you test the man/troff output? Wouldn't it make sense to also have it in the testsuite?

File: ocamldoc/odoc_sig.ml [hunk 3]
```diff
@@ -183,15 +203,13 @@ module Analyser =
                 assert false
 
               | (name, _atts, ct) :: [] ->
-                let pos = ct.Parsetree.ptyp_loc.Location.loc_end.Lexing.pos_cnum in
-                let s = get_string_of_file pos pos_end in
-                let (_,comment_opt) =  My_ir.just_after_special !file_name s in
+                  let pos = Loc.ptyp_end ct in
+                  let (_,comment_opt) = just_after_special pos pos_end in
```
Comment (gasche) [inline ocamldoc/odoc_sig.ml]: nitpicking: why change indentation here?

**Comment by gasche** [comment]
> Note that the first commit in this PR is a questionable refactoring commit that I implemented when trying to decrease my perception of the signal to noise ratio in odoc_sig.ml, it can be reverted.

I think the refactoring is nice. If you prefer to have it as a first PR, we can do it this way -- it can be merged right away -- but otherwise it can just stay in the present PR.

File: testsuite/tests/tool-ocamldoc-2/inline_records.mli (hunk 1 above)
Comment (Octachron) [inline testsuite/tests/tool-ocamldoc-2/inline_records.mli:26]: It is a very good point, I added such case.

File: testsuite/tests/tool-ocamldoc-2/inline_records.reference (hunk 2 above)
Comment (Octachron) [inline testsuite/tests/tool-ocamldoc-2/inline_records.reference:102]: It was an error due to some problem with execution order within `odoc_latex`. The field documentation was originally preprocessed but only in text mode (i.e. without some of the fancy latex feature). However, I found the time to rewrite the code to support a full latex mode for field documentations.

**Comment by Octachron** [comment]
I have added a test for the troff output, and improved the field documentation support for all three generators (superseding the fix for the latex output along the way). There is still some room for some improvements: currently, inline records are rendered in the same way as normal record which is probably not optimal (but better than using the simple text renderer as with the previous version of the PR).

File: ocamldoc/odoc_sig.ml (hunk 3 above)
Comment (Octachron) [inline ocamldoc/odoc_sig.ml]: Good question? Fixed.

**Comment by gasche** [comment]
I would be glad to merge, but it would be nice if you could rebase to squash the extraneous commits.

**Comment by Octachron** [comment]
Done. I have also updated the wording of the commit and Changes entry since the inline record support is not so light anymore.

**Comment by gasche** [comment]
Many thanks for your continued ocamldoc contributions.

**Pull request merged**